ui/raidboss/popup-text.ts

```diff
@@ -2,14 +2,86 @@ import { callOverlayHandler, addOverlayListener } from '../../resources/overlay_
 
 import AutoplayHelper from './autoplay_helper';
 import BrowserTTSEngine from './browser_tts_engine';
-import { addPlayerChangedOverrideListener } from '../../resources/player_override';
+import { addPlayerChangedOverrideListener, PlayerChangedDetail } from '../../resources/player_override';
 import PartyTracker from '../../resources/party';
 import Regexes from '../../resources/regexes';
 import Util from '../../resources/util';
 import ZoneId from '../../resources/zone_id';
+import {
+  LooseTrigger, OutputStrings, LocaleText, TriggerSet, TimelineFunc, LooseTriggerSet,
+  ResponseField, TriggerAutoConfig, MatchesAny, TriggerField, TriggerOutput,
+  TriggerFunc, OptionalUnlessVoid, Output, ResponseOutput, NetRegexTrigger, RegexTrigger,
+} from '../../types/trigger';
+import { UnreachableCode } from '../../resources/not_reached';
+import { Lang, NonEnLang } from '../../resources/languages';
+import { PerTriggerAutoConfig, PerTriggerOption, RaidbossOptions } from './raidboss_options';
+import { TimelineReplacement, TimelineLoader } from './timeline';
+import { RaidbossFileData } from './data/raidboss_manifest.txt';
+import { RaidbossData } from '../../types/data';
+import { Job, Role } from '../../types/job';
+import { EventResponses, LogEvent } from '../../types/event';
+
+export const isNetRegexTrigger =
+  (trigger?: LooseTrigger): trigger is NetRegexTrigger<RaidbossData> => {
+    if (trigger)
+      return 'netRegex' in trigger;
+    return false;
+  };
+
+export const isRegexTrigger = (trigger?: LooseTrigger): trigger is RegexTrigger<RaidbossData> => {
+  if (trigger)
+    return 'regex' in trigger;
+  return false;
+};
+
+// @TODO: Maybe there's a better way to define this?
+type RegexParserLang = `regex${Capitalize<Lowercase<NonEnLang>>}` | 'regex';
+type NetRegexParserLang = `netRegex${Capitalize<Lowercase<NonEnLang>>}` | 'netRegex';
+
+export type RaidbossLooseTrigger = LooseTrigger & {
+  filename?: string;
+  localRegex?: RegExp;
+  localNetRegex?: RegExp;
+  output?: Output;
+};
+
+export type RaidbossLooseTimelineTrigger = RaidbossLooseTrigger & {
+  isTimelineTrigger?: boolean;
+};
+
+export type RaidbossTriggerSet = LooseTriggerSet & {
+  filename?: string;
+  timelineTriggers?: RaidbossLooseTimelineTrigger[];
+  triggers?: RaidbossLooseTrigger[];
+};
+
+const isRaidbossLooseTimelineTrigger =
+  (trigger: LooseTrigger): trigger is RaidbossLooseTimelineTrigger => {
+    return 'isTimelineTrigger' in trigger;
+  };
+
+// Due to the nature of OutputStrings, there's no simple way to check that an object is actually
```

Comment:
I was trying to suggest the following, treating dynamic responseOutputStrings as potentially unknown objects and checking them as we use them, rather than trying to figure out if they are the right type up front.  I think this is simpler, but maybe you disagree?

```diff
diff --git a/ui/raidboss/popup-text.ts b/ui/raidboss/popup-text.ts
index 61949018..e780f141 100644
--- a/ui/raidboss/popup-text.ts
+++ b/ui/raidboss/popup-text.ts
@@ -59,23 +59,6 @@ const isRaidbossLooseTimelineTrigger =
     return 'isTimelineTrigger' in trigger;
   };
 
-// Due to the nature of OutputStrings, there's no simple way to check that an object is actually
-// correct. Have to loop over objects to ensure they have the proper structure
-const isOutputStrings = (check: unknown): check is OutputStrings => {
-  if (!(check instanceof Object))
-    return false;
-  for (const propCheck of Object.values(check)) {
-    if (!(propCheck instanceof Object))
-      return false;
-    const prop = propCheck as unknown;
-    if (!(prop instanceof Object))
-      return false;
-    if (!('en' in prop))
-      return false;
-  }
-  return true;
-};
-
 // There should be (at most) six lines of instructions.
 const raidbossInstructions: { [lang in Lang]: string[] } = {
   en: [
@@ -242,12 +225,14 @@ class OrderedTriggerList {
   }
 }
 
+const isObject = (x: unknown): x is { [key: string]: unknown } => typeof x === 'object';
+
 type TriggerParams = { [key: string]: string };
 
 class TriggerOutputProxy {
   public outputStrings: OutputStrings;
   public overrideStrings: OutputStrings = {};
-  public responseOutputStrings: OutputStrings = {};
+  public responseOutputStrings: { [outputName: string]: unknown } = {};
   public unknownValue = '???';
   constructor(
       public trigger: RaidbossLooseTrigger,
@@ -283,8 +268,8 @@ class TriggerOutputProxy {
       // the config ui, the response should return the exact same set of
       // outputStrings every time.  Thank you for coming to my TED talk.
       set(target, property, value): boolean {
-        if (property === 'responseOutputStrings' && isOutputStrings(value)) {
-          target[property] = value;
+        if (property === 'responseOutputStrings' && isObject(value)) {
+          target.responseOutputStrings = value;
           return true;
         }
 
@@ -309,8 +294,11 @@ class TriggerOutputProxy {
           // or there's no response and trigger provides everything.  Having
           // this well-defined smooths out the collision edge cases.
           let str = target.getReplacement(target.overrideStrings[name], params, name, id);
-          if (str === undefined)
-            str = target.getReplacement(target.responseOutputStrings[name], params, name, id);
+          if (str === undefined) {
+            const responseString = target.responseOutputStrings[name];
+            if (isObject(responseString))
+              str = target.getReplacement(responseString, params, name, id);
+          }
           if (str === undefined)
             str = target.getReplacement(target.outputStrings[name], params, name, id);
           if (str === undefined) {
@@ -325,7 +313,7 @@ class TriggerOutputProxy {
 
   getReplacement(
       // Can't use optional modifier for this arg since the others aren't optional
-      template: LocaleText | undefined,
+      template: { [lang: string]: unknown } | undefined,
       params: TriggerParams,
       name: string,
       id: string): string | undefined {
```

Reply:
I disagree with this approach because the check would have to be repeated for each of the three `outputStrings` checks in `TriggerOutputProxy`'s `get`'s calls to `target.getReplacement`.

I'd rather do the validation in the constructor once rather than have to check all of these every time the trigger runs, and I'd rather use a helper function to verify than repeat the logic every time.

Reply:
I am not sure I follow.  This check doesn't have to be repeated, it is just the once for the `responseOutputStrings` field.

Reply:
Sorry, I got this confused because your suggested change has the check in the `get`, so my mind just jumped to "well we should check the other two as well". I still think it doesn't make sense to check this in the `get`, I think the check should be in the `set` so it's only run once?

Reply:
This is adjusted again with the latest commit, hopefully that plus my comment above help clear up my intent here?

Reply:
Alright, take 4. Kept what you originally suggested but added an error message to the `set` for better feedback.

Reply:
Thanks for the extra error message!  Sorry to be hardheaded about this.  <_<

Reply:
I don't think you were being hardheaded about this, I think we just had different thoughts about the best way to deal with it and I wasn't articulating my point of 'we should really be checking this on set' very well. 😄